src/core/lib/gprpp/sync.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Inline implementation of synchronization primitives, preferred over the C
+   public API for C++ code. */
+
+#ifndef GRPC_CORE_LIB_GPRPP_SYNC_H
+#define GRPC_CORE_LIB_GPRPP_SYNC_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <assert.h>
+#include <stdint.h>
+#include <atomic>
+#include <type_traits>
+
+#include "src/core/lib/gprpp/memory.h"
+
+namespace grpc_core {
+
+class RefCount {
+ public:
+  explicit RefCount(intptr_t init = 1) : val(init) {}
+
+  // Reinitializes the ref-count.
+  void Reset(intptr_t init) { val.store(init, std::memory_order_relaxed); }
+
+  void Ref(intptr_t n = 1) { val.fetch_add(n, std::memory_order_relaxed); }
+
+  void RefNonZero() {
+#ifndef NDEBUG
+    const intptr_t prior = val.fetch_add(1, std::memory_order_relaxed);
+    assert(prior > 0);
+#else
+    Ref();
+#endif
+  }
+
+  bool RefIfNonZero() {
```

Comment:
What's the expected use-case for this method?  My gut reaction is that if we actually need to use this, then it's probably a sign that the calling code isn't structured the right way.

Reply:
Sorry if these are not clear. The code was copy/pasted from a larger WIP PR. This is the main user of this method:

https://github.com/grpc/grpc/blob/master/src/core/ext/transport/chttp2/transport/writing.cc#L165

The implementation here is actually 3x faster due to only have one locked atomic op. :-)